draft/ZEP0010.md

```diff
@@ -0,0 +1,413 @@
+- [Josh Moore](https://github.com/joshmoore), German BioImaging
+
+Status: Draft
+
+Type: Specification
+
+Created: 2025-05-12
+
+## Abstract
+
+This proposal defines a new generic extension point, ``extensions``, to be
+included in the metadata of Zarr v3 arrays and groups. The ``extensions`` field
+provides a consistent mechanism for attaching additional metadata that does not
+fit into existing extension points defined by the core specification. Extension
+entries within this field follow the naming and structure rules established in
+ZEP0009. This mechanism enables third parties to define and share metadata
+extensions without requiring changes to the core specification or introducing
+new top-level keys.
+
+## Introduction
+
+Zarr specification version 3 currently defines four extension points, each
+associated with a specific (array) metadata field. Additional extension points
+may be added by future ZEPs. Until that time, however, third-parties may want
+to add arbitrary extension objects to either arrays or groups. This proposal
+introduces a generic ``extensions`` field that serves as a container for such a
+list of extensions.
+
+## Proposal
+
+To provide for more flexible, immediate, and de-centralized use cases, we
+propose to add a generic extension point ```extensions`` on
+both arrays and groups into which extensions MAY be added.
+
+The ``extensions`` field if present MUST contain an array of extension
+definitions. The contained array MUST either have one or more extensions or the
+object MUST be omitted entirely.
+
+Further details on the specification changes can be found in
+<https://github.com/zarr-developers/zarr-specs/pull/344>.
+
+### Definition and naming
+
+Each extension object will follow the rules laid out in the "Zarr extensions"
+section of the v3 specification.
+
+### Processing
+
+In a manner similar to OpenGL’s extension mechanism, where implementations must
+explicitly advertise and support extensions they recognize, Zarr implementers
+are expected to inspect the extensions array and determine whether each listed
+extension is supported. If an extension includes ``"must_understand": true``
+(the default) and the implementation does not support it, the dataset must not
+be loaded and an appropriate error should be raised. For extensions with
+``"must_understand": false``, implementers may safely ignore unrecognized entries.
+
+To support a given extension, an implementation must either (1) check for known
+extension names and invoke appropriate logic according to the extension’s
+specification at the correct point in its processing pipeline (e.g., during
+metadata interpretation, data access, or layout resolution), or (2) delegate
+that logic via a callback or plugin mechanism that allows third-party code to
+handle the extension dynamically. This modular approach enables implementers to
+support a flexible and evolving set of extensions while maintaining core
+compatibility.
```

Comment:
these two options boil down to "use your code" or "use third party code", which don't seem that different to me. Without constraints on the scope of these extensions, there can't really be a specific definition of what it necessary to support them, so maybe this section can be omitted, or replaced with a statement to the effect that each extension will define the constraints required for supporting it.

Reply:
This section could be omitted. It is not normative, because implementations can handle this as they please. It might be useful to present a range of options to developers, though.